subsys/nfc/t4t/hl_procedure.c

```diff
@@ -0,0 +1,476 @@
+
+		return t4t_hl_data_exchange(&apdu_comm);
+	}
+
+	file_id = sys_get_be16(t4t_hl.ndef.file_id);
+
+	err = t4t_file_assign(file_id);
+	if (err) {
+		return err;
+	}
+
+	if (hl_cb->ndef_read) {
+		hl_cb->ndef_read(file_id, t4t_hl.ndef.buff,
+				 t4t_hl.ndef.nlen + NDEF_FILE_NLEN_SIZE);
+	}
+
+	return 0;
+}
+
+static int ndef_file_chunk_update(void)
+{
+	struct nfc_t4t_apdu_comm apdu_comm;
+	u8_t nlen_data[NDEF_FILE_NLEN_SIZE];
+
+	nfc_t4t_apdu_comm_clear(&apdu_comm);
+
+	apdu_comm.instruction = NFC_T4T_APDU_COMM_INS_UPDATE;
+
+	if (t4t_hl.file_offset < t4t_hl.ndef.buff_size) {
+		apdu_comm.parameter = t4t_hl.file_offset;
+		apdu_comm.data.buff = t4t_hl.ndef.buff + t4t_hl.file_offset;
+		apdu_comm.data.len = MIN(t4t_hl.ndef.buff_size - t4t_hl.file_offset,
+				MIN(APDU_LE_MAP_2_MAX_VALUE, t4t_hl.ndef.cc->max_capdu_size));
+
+		t4t_hl.file_offset += apdu_comm.data.len;
+		t4t_hl.transaction_type = NFC_T4T_HL_NDEF_UPDATE;
+
+		return t4t_hl_data_exchange(&apdu_comm);
+	}
+
+	sys_put_be16(t4t_hl.ndef.nlen, nlen_data);
+
+	apdu_comm.parameter = 0;
+	apdu_comm.data.buff = nlen_data;
+	apdu_comm.data.len = sizeof(nlen_data);
+
+	t4t_hl.transaction_type = NFC_T4T_HL_NDEF_NLEN_UPDATE;
+
+	return t4t_hl_data_exchange(&apdu_comm);
+}
+
+static void on_ndef_nlen_update(void)
+{
+	u16_t file_id = sys_get_be16(t4t_hl.ndef.file_id);
+
+	if (hl_cb->ndef_updated) {
+		hl_cb->ndef_updated(file_id);
+	}
+}
+
+static int on_rapdu_succes(const struct nfc_t4t_apdu_resp *resp)
+{
+	int err = 0;
+
+	switch (t4t_hl.transaction_type) {
+	case NFC_T4T_HL_SELECT:
+		if (hl_cb->selected) {
+			hl_cb->selected(t4t_hl.select_type);
+		}
+
+		break;
+
+	case NFC_T4T_HL_CC_READ:
+		err = on_cc_read(resp);
+		break;
+
+	case NFC_T4T_HL_NDEF_NLEN_READ:
+		err = on_ndef_nlen_read(resp);
+		if (err) {
+			LOG_ERR("NDEF NLEN field read failed.");
+			break;
+		}
+
+		err = ndef_file_chunk_read(resp);
+
+		break;
+
+	case NFC_T4T_HL_NDEF_READ:
+		err = ndef_file_chunk_read(resp);
+
+		break;
+
+	case NFC_T4T_HL_NDEF_NLEN_CLEAR:
+	case NFC_T4T_HL_NDEF_UPDATE:
+		err = ndef_file_chunk_update();
+
+		break;
+
+	case NFC_T4T_HL_NDEF_NLEN_UPDATE:
+		on_ndef_nlen_update();
+
+		break;
+
+	default:
+		err = -EAGAIN;
+
+	}
+
+	return err;
+}
+
+int nfc_t4t_hl_procedure_cb_register(const struct nfc_t4t_hl_procedure_cb *cb)
+{
+	if (!cb) {
+		return -EINVAL;
+	}
+
+	hl_cb = cb;
+
+	return 0;
+}
+
+int nfc_t4t_hl_procedure_on_data_received(const u8_t *data, size_t len)
+{
+	int err = 0;
+	struct nfc_t4t_apdu_resp apdu_resp;
+
+	if ((!data) || (len < RAPDU_MIN_LEN)) {
+		return -EINVAL;
+	}
+
+	nfc_t4t_apdu_resp_clear(&apdu_resp);
+
+	err = nfc_t4t_apdu_resp_decode(&apdu_resp, data, len);
+	if (err) {
+		LOG_ERR("NFC T4T R-APDU decode error: %d", err);
+		return err;
+	}
+
+	nfc_t4t_apdu_resp_printout(&apdu_resp);
+
+	if (apdu_resp.status == NFC_T4T_APDU_RAPDU_STATUS_CMD_COMPLETED) {
+		err = on_rapdu_succes(&apdu_resp);
+	} else {
+		LOG_ERR("R-APDU File not found. No data returned.");
```

Comment:
RAPDU can return other status errors (not only file not found). Please use more general log and print the status value:

R-APDU returned error status: ..

Reply:
Ok LOG is more generic now.